pkg/terra/median_contract.go

```diff
@@ -29,18 +31,92 @@ type LatestConfigReader interface {
 }
 
 type MedianContract struct {
+	utils.StartStopOnce
 	address     sdk.AccAddress
 	chainReader client.Reader
 	lggr        Logger
 	cr          LatestConfigReader
 	cfg         Config
+	stop, done  chan struct{}
+
+	// cached state
+	mu              sync.RWMutex
+	ts              time.Time
+	configDigest    types.ConfigDigest
+	epoch           uint32
+	round           uint8
+	latestAnswer    *big.Int
+	latestTimestamp time.Time
 }
 
 func NewMedianContract(address sdk.AccAddress, chainReader client.Reader, lggr Logger, cr LatestConfigReader, cfg Config) *MedianContract {
-	return &MedianContract{address: address, chainReader: chainReader, lggr: lggr, cr: cr, cfg: cfg}
+	return &MedianContract{
+		address:     address,
+		chainReader: chainReader,
+		lggr:        lggr,
+		cr:          cr,
+		cfg:         cfg,
+		stop:        make(chan struct{}),
+		done:        make(chan struct{}),
+	}
+}
+
+func (ct *MedianContract) Start() error {
+	return ct.StartOnce("TerraMedianContract", func() error {
+		ct.lggr.Debugf("Starting")
+		go ct.pollState()
+		return nil
+	})
+}
+
+func (ct *MedianContract) Close() error {
+	return ct.StopOnce("TerraMedianContract", func() error {
+		ct.lggr.Debugf("Stopping")
+		close(ct.stop)
+		<-ct.done
+		return nil
+	})
+}
+
+func (ct *MedianContract) pollState() {
```

Comment:
yeah I think it would be cleaner to combine these pollers into a single state cache object like you mentioned. It would get injected into MedianContract/ContractTracker/ContractTransmitter etc. who accept an interface describing whatever they need from the cache. Just a single goroutine + for/select loop would be make it easier to reason about the rpc call frequency. Wdyt?

Reply:
Second this!